### PR TITLE
Instrument proctelemetry.ProcessMetrics

### PR DESCRIPTION
**Pull request by moh-osman3**
Description:

This PR instruments proctelemetry.ProcessMetrics with otel go. This is a followup PR that is based on similar changes made for obsreport.Receiver: https://github.com/open-telemetry/opentelemetry-collector/pull/6222

Link to tracking Issue: Part of https://github.com/open-telemetry/opentelemetry-collector/issues/816

Testing:
Built otelcorecol binary and checked metrics endpoint. Tested with and without featuregate.

```
% curl localhost:8888/metrics
# HELP otelcol_process_cpu_seconds_total Total CPU user and system time in seconds
# TYPE otelcol_process_cpu_seconds_total counter
otelcol_process_cpu_seconds_total 0.19
# HELP otelcol_process_memory_rss Total physical memory (resident set size)
# TYPE otelcol_process_memory_rss gauge
otelcol_process_memory_rss 3.467264e+07
# HELP otelcol_process_runtime_heap_alloc_bytes Bytes of allocated heap objects (see 'go doc runtime.MemStats.HeapAlloc')
# TYPE otelcol_process_runtime_heap_alloc_bytes gauge
otelcol_process_runtime_heap_alloc_bytes 5.4080584e+08
# HELP otelcol_process_runtime_total_alloc_bytes_total Cumulative bytes allocated for heap objects (see 'go doc runtime.MemStats.TotalAlloc')
# TYPE otelcol_process_runtime_total_alloc_bytes_total counter
otelcol_process_runtime_total_alloc_bytes_total 5.43115264e+08
# HELP otelcol_process_runtime_total_sys_memory_bytes Total bytes of memory obtained from the OS (see 'go doc runtime.MemStats.Sys')
# TYPE otelcol_process_runtime_total_sys_memory_bytes gauge
otelcol_process_runtime_total_sys_memory_bytes 5.78296976e+08
# HELP otelcol_process_uptime_total Uptime of the process
# TYPE otelcol_process_uptime_total counter
otelcol_process_uptime_total 104.306318
# HELP otelcol_target_info Target metadata
# TYPE otelcol_target_info gauge
otelcol_target_info{service_instance_id="26829589-afdd-4f5a-b201-b9339e01ff80",service_name="otelcorecol",service_version="0.68.0-dev"} 1
```

### CODE REVIEW

**Comment by moh-osman3** [comment]
@jpkrohling @bogdandrutu @paivagustavo This is my latest instrumentation PR. It should be ready for review if y'all have any feedback! Thank you!

**Comment by jmacd** [comment]
Looks good to me. FWIW eventually I would like to see this stuff replaced with an otel-go metrics instrumentation library, for example here https://github.com/lightstep/otel-launcher-go/tree/main/lightstep/instrumentation/runtime.

File: service/service_test.go [hunk 1]
```diff
@@ -226,7 +226,7 @@ func TestServiceTelemetryWithOpenTelemetryMetrics(t *testing.T) {
 			registry := featuregate.NewRegistry()
 			obsreportconfig.RegisterInternalMetricFeatureGate(registry)
 			colTel := newColTelemetry(registry)
-			require.NoError(t, colTel.registry.Apply(map[string]bool{obsreportconfig.UseOtelForInternalMetricsfeatureGateID: true}))
+			require.NoError(t, colTel.registry.Apply(map[string]bool{obsreportconfig.UseOtelForInternalMetricsfeatureGateID: false}))
```
Comment (codeboten) [inline service/service_test.go:229]: Why is this changed here?

File: service/internal/proctelemetry/process_telemetry.go [hunk 2]
```diff
@@ -46,19 +71,33 @@ type processMetrics struct {
...
 	}
-	var err error
 	pm.proc, err = process.NewProcess(int32(os.Getpid()))
 	if err != nil {
 		return err
 	}
 
-	pm.processUptime, err = registry.AddFloat64DerivedCumulative(
+	if pm.useOtelForMetrics {
+		return pm.recordWithOtel()
```
Comment (codeboten) [inline service/internal/proctelemetry/process_telemetry.go:92]: I would move line 84 in this same if here

File: service/service_test.go (hunk 1 above)
Comment (moh-osman3) [reply service/service_test.go]: Thanks for the review! Not sure why I left this file like this. I think I noticed `assertMetrics` checks for labels in every metric that doesn't apply when the featuregate is enabled. So this test will fail for otel. Instead of just flipping the featuregate off I am now checking if the featuregate is disabled before calling `assertMetrics`. I am definitely open to suggestions for fixing/improving this test.

File: service/internal/proctelemetry/process_telemetry.go (hunk 2 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry.go:90]: Fixed!

File: service/internal/proctelemetry/process_telemetry_test.go [hunk 3]
```diff
@@ -36,14 +62,99 @@ var expectedMetrics = []string{
...
+
+	for _, metricName := range tel.expectedMetrics {
+		metric, ok := mp[metricName]
+		if !ok {
+			withSuffix := metricName + "_total"
+			metric, ok = mp[withSuffix]
+		}
+		require.True(t, ok)
+		require.True(t, len(metric.Metric) == 1)
+		// require.True(t, metric.GetType() == io_prometheus_client.MetricType_COUNTER)
```
Comment (codeboten) [inline service/internal/proctelemetry/process_telemetry_test.go:132]: Should this be removed?

File: service/internal/proctelemetry/process_telemetry_test.go [hunk 4]
```diff
@@ -36,14 +62,99 @@ var expectedMetrics = []string{
...
+		var metricValue float64
+		if metric.GetType() == io_prometheus_client.MetricType_COUNTER {
+			metricValue = metric.Metric[0].GetCounter().GetValue()
+		} else {
+			metricValue = metric.Metric[0].GetGauge().GetValue()
+		}
+		if strings.HasPrefix(metricName, "process_uptime") || strings.HasPrefix(metricName, "process_cpu_seconds") {
+			// This likely will still be zero when running the test.
+			assert.True(t, metricValue >= 0, metricName)
```
Comment (codeboten) [inline service/internal/proctelemetry/process_telemetry_test.go:141]:
```suggestion
			assert.GreaterOrEqual(t, metricValue, 0, metricName)
```

File: service/internal/proctelemetry/process_telemetry_test.go [hunk 5]
```diff
@@ -36,14 +62,99 @@ var expectedMetrics = []string{
...
+			continue
+		}
+
+		assert.True(t, metricValue > 0, metricName)
```
Comment (codeboten) [inline service/internal/proctelemetry/process_telemetry_test.go:145]:
```suggestion
		assert.Greater(t, metricValue, 0, metricName)
```

File: service/service_test.go [hunk 6]
```diff
@@ -269,7 +269,9 @@ func testCollectorStartHelper(t *testing.T, useOtel bool, tc ownMetricsTestCase)
 		// Sleep for 1 second to ensure the http server is started.
 		time.Sleep(1 * time.Second)
 		assert.True(t, loggingHookCalled)
-		assertMetrics(t, metricsAddr, tc.expectedLabels)
+		if !useOtel {
```
Comment (codeboten) [inline service/service_test.go:272]: Should assertMetrics not work regardless of whether it's using otel or not?

Comment (moh-osman3) [reply service/service_test.go:272]: If I'm understanding `assertMetrics` correctly, it is meant to check that certain labels are present from the metrics found at the metrics endpoint.  Now that otel-go instrumentation is used when featuregate is enabled, these labels do not apply to otel.  OCMetrics however remain unchanged, so they will have the expected labels.

Comment (moh-osman3) [reply service/service_test.go:272]: For reference:
*featuregate enabled* only a few of the metrics have the expected labels like `service_instance_id`
```
% curl localhost:8888/metrics                         
# HELP otelcol_exporter_enqueue_failed_log_records Number of log records failed to be added to the sending queue.
# TYPE otelcol_exporter_enqueue_failed_log_records counter
otelcol_exporter_enqueue_failed_log_records{exporter="logging",service_instance_id="54494d1d-76a1-42e2-a128-9bdbb0668396",service_name="otelcorecol",service_version="0.68.0-dev"} 0
# HELP otelcol_exporter_enqueue_failed_metric_points Number of metric points failed to be added to the sending queue.
# TYPE otelcol_exporter_enqueue_failed_metric_points counter
otelcol_exporter_enqueue_failed_metric_points{exporter="logging",service_instance_id="54494d1d-76a1-42e2-a128-9bdbb0668396",service_name="otelcorecol",service_version="0.68.0-dev"} 0
# HELP otelcol_exporter_enqueue_failed_spans Number of spans failed to be added to the sending queue.
# TYPE otelcol_exporter_enqueue_failed_spans counter
otelcol_exporter_enqueue_failed_spans{exporter="logging",service_instance_id="54494d1d-76a1-42e2-a128-9bdbb0668396",service_name="otelcorecol",service_version="0.68.0-dev"} 0
# HELP otelcol_process_cpu_seconds_total Total CPU user and system time in seconds
# TYPE otelcol_process_cpu_seconds_total counter
otelcol_process_cpu_seconds_total 0.06
# HELP otelcol_process_memory_rss Total physical memory (resident set size)
# TYPE otelcol_process_memory_rss gauge
otelcol_process_memory_rss 3.432448e+07
# HELP otelcol_process_runtime_heap_alloc_bytes Bytes of allocated heap objects (see 'go doc runtime.MemStats.HeapAlloc')
# TYPE otelcol_process_runtime_heap_alloc_bytes gauge
otelcol_process_runtime_heap_alloc_bytes 5.4054712e+08
# HELP otelcol_process_runtime_total_alloc_bytes_total Cumulative bytes allocated for heap objects (see 'go doc runtime.MemStats.TotalAlloc')
# TYPE otelcol_process_runtime_total_alloc_bytes_total counter
otelcol_process_runtime_total_alloc_bytes_total 5.42846208e+08
# HELP otelcol_process_runtime_total_sys_memory_bytes Total bytes of memory obtained from the OS (see 'go doc runtime.MemStats.Sys')
# TYPE otelcol_process_runtime_total_sys_memory_bytes gauge
otelcol_process_runtime_total_sys_memory_bytes 5.77772688e+08
# HELP otelcol_process_uptime_total Uptime of the process
# TYPE otelcol_process_uptime_total counter
otelcol_process_uptime_total 10.014383
# HELP otelcol_target_info Target metadata
# TYPE otelcol_target_info gauge
otelcol_target_info{service_instance_id="54494d1d-76a1-42e2-a128-9bdbb0668396",service_name="otelcorecol",service_version="0.68.0-dev"} 1
```

Comment (moh-osman3) [reply service/service_test.go:272]: *featuregate disabled* (OC) all the metrics seem to have the labels that assertMetrics is expecting
```
% curl localhost:8888/metrics   
# HELP otelcol_exporter_enqueue_failed_log_records Number of log records failed to be added to the sending queue.
# TYPE otelcol_exporter_enqueue_failed_log_records counter
otelcol_exporter_enqueue_failed_log_records{exporter="logging",service_instance_id="d86342f8-6b14-4451-afdc-9fcd60b7aee8",service_name="otelcorecol",service_version="0.68.0-dev"} 0
# HELP otelcol_exporter_enqueue_failed_metric_points Number of metric points failed to be added to the sending queue.
# TYPE otelcol_exporter_enqueue_failed_metric_points counter
otelcol_exporter_enqueue_failed_metric_points{exporter="logging",service_instance_id="d86342f8-6b14-4451-afdc-9fcd60b7aee8",service_name="otelcorecol",service_version="0.68.0-dev"} 0
# HELP otelcol_exporter_enqueue_failed_spans Number of spans failed to be added to the sending queue.
# TYPE otelcol_exporter_enqueue_failed_spans counter
otelcol_exporter_enqueue_failed_spans{exporter="logging",service_instance_id="d86342f8-6b14-4451-afdc-9fcd60b7aee8",service_name="otelcorecol",service_version="0.68.0-dev"} 0
# HELP otelcol_process_cpu_seconds Total CPU user and system time in seconds
# TYPE otelcol_process_cpu_seconds counter
otelcol_process_cpu_seconds{service_instance_id="d86342f8-6b14-4451-afdc-9fcd60b7aee8",service_name="otelcorecol",service_version="0.68.0-dev"} 0.08
# HELP otelcol_process_memory_rss Total physical memory (resident set size)
# TYPE otelcol_process_memory_rss gauge
otelcol_process_memory_rss{service_instance_id="d86342f8-6b14-4451-afdc-9fcd60b7aee8",service_name="otelcorecol",service_version="0.68.0-dev"} 3.426304e+07
# HELP otelcol_process_runtime_heap_alloc_bytes Bytes of allocated heap objects (see 'go doc runtime.MemStats.HeapAlloc')
# TYPE otelcol_process_runtime_heap_alloc_bytes gauge
otelcol_process_runtime_heap_alloc_bytes{service_instance_id="d86342f8-6b14-4451-afdc-9fcd60b7aee8",service_name="otelcorecol",service_version="0.68.0-dev"} 5.4050804e+08
# HELP otelcol_process_runtime_total_alloc_bytes Cumulative bytes allocated for heap objects (see 'go doc runtime.MemStats.TotalAlloc')
# TYPE otelcol_process_runtime_total_alloc_bytes counter
otelcol_process_runtime_total_alloc_bytes{service_instance_id="d86342f8-6b14-4451-afdc-9fcd60b7aee8",service_name="otelcorecol",service_version="0.68.0-dev"} 5.4280048e+08
# HELP otelcol_process_runtime_total_sys_memory_bytes Total bytes of memory obtained from the OS (see 'go doc runtime.MemStats.Sys')
# TYPE otelcol_process_runtime_total_sys_memory_bytes gauge
otelcol_process_runtime_total_sys_memory_bytes{service_instance_id="d86342f8-6b14-4451-afdc-9fcd60b7aee8",service_name="otelcorecol",service_version="0.68.0-dev"} 5.77772688e+08
# HELP otelcol_process_uptime Uptime of the process
# TYPE otelcol_process_uptime counter
otelcol_process_uptime{service_instance_id="d86342f8-6b14-4451-afdc-9fcd60b7aee8",service_name="otelcorecol",service_version="0.68.0-dev"} 3.199575
```

Comment (moh-osman3) [reply service/service_test.go:272]: Do you think I should manually enrich the proctelemetry.ProcessMetrics with these `service_*` labels by adding the otelattrs? Is there a way to retrieve the values of these labels to add them to otel, or should these not apply anymore?

File: service/internal/proctelemetry/process_telemetry_test.go (hunk 3 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry_test.go]: Yes thanks for catching that! metrics can be both gauge or counter so this assert doesn't make sense

File: service/internal/proctelemetry/process_telemetry_test.go (hunk 4 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry_test.go]: Thanks for the suggestion - for consistency, I also updated the OC unit test to use the `Greater` functions

File: service/internal/proctelemetry/process_telemetry_test.go (hunk 5 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry_test.go]: Fixed!

File: service/service_test.go (hunk 6 above)
Comment (codeboten) [reply service/service_test.go:272]: I would expect those labels to process metrics as they do for exporter metric, see `otelcol_exporter_enqueue_failed_log_records` for example.

Comment (moh-osman3) [reply service/service_test.go:272]: @codeboten Hmm but isn't `otelcol_exporter_enqueue_failed_log_records` an OC instrument and not an Otel one?  I think so based off this [file](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/obsreport.go#L78).  So I would expect it to have the `service_*` labels.  I don't think otel provides these same labels. Is there any way we can confirm this?

Comment (moh-osman3) [reply service/service_test.go:272]: @codeboten So after further discussion with Jmac,  it seems resource attributes are created during [telemetry initialization](https://github.com/open-telemetry/opentelemetry-collector/blob/e5208293ec5d4e04939ff52d60519ddbaa12d87a/service/telemetry.go#L93) and passed to both OC and Otel initialization functions.

This helped realize that these attributes actually aren't missing at all, but instead recorded in one place for otel - the `target_info` metric.  According to otel specification about resource attributes 

```
In SDK Prometheus (pull) exporters, resource attributes SHOULD be converted to a single target_info metric; otherwise, they MUST be dropped, and MUST NOT be attached as labels to other metric families.
```

So the resource attributes are simply captured in the `otel_target_info` resource attributes (it is not necessary to have these attributes repeat in the other metrics). So I think the sdk exposes these attributes once in the target_info metric through the prometheus exporter. And therefore these attributes can be recovered for all the otel metrics at query-time from this `target_info`.  

cc @jmacd in case I'm missing any important info

Comment (jmacd) [reply service/service_test.go:272]: LGTM thanks @moh-osman3

Comment (codeboten) [reply service/service_test.go:272]: Thanks @moh-osman3, now that I look at `assertMetrics` more closely I realize that my confusion stemmed from thinking this was performing the same type of test as the `assertMetrics` in the batchprocessor tests was doing. Seeing the output above with metrics emitted from both OC and OT further confused me. Thank you for clarifying. 

A follow up to this would be to change `assertMetrics` to either validate the metrics themselves, or rename it to `assertOCLabels` (or something like that).

**Review by codeboten — APPROVED** [review]

File: service/service_test.go (hunk 6 above)
Comment (moh-osman3) [reply service/service_test.go:272]: Thanks for the review @codeboten !  Your questions were helpful to improve my understanding as well. I created this issue https://github.com/open-telemetry/opentelemetry-collector/issues/7075 to track your suggested improvements on the `assertMetrics` function.

**Review by bogdandrutu — COMMENTED** [review]
This needs updates because of API update in otel.

**Review by bogdandrutu — CHANGES_REQUESTED** [review]
Please rebase, will need API fixes.

**Comment by moh-osman3** [comment]
> Please rebase, will need API fixes.

bogdandrutu Thanks for pointing this out! I just updated based on the otel api changes. Can you please take another look?

**Comment by moh-osman3** [comment]
> Please rebase, will need API fixes.

@bogdandrutu Thanks for pointing this out. I just updated based on the otel api changes. Can you please take another look?

File: .chloggen/proctelemetry-processmetrics.yaml [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. otlpreceiver)
+component: proctelemetry
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "Instrument `proctelemetry.ProcessMetrics` metrics with otel-go"
+
+# One or more tracking issues or pull requests related to the change
+issues: [6886]
```
Comment (bogdandrutu) [inline .chloggen/proctelemetry-processmetrics.yaml:11]: New line.

File: service/internal/proctelemetry/process_telemetry.go [hunk 8]
```diff
@@ -46,19 +66,31 @@ type processMetrics struct {
...
-	var err error
+
 	pm.proc, err = process.NewProcess(int32(os.Getpid()))
 	if err != nil {
 		return err
 	}
 
-	pm.processUptime, err = registry.AddFloat64DerivedCumulative(
+	if pm.useOtelForMetrics {
+		pm.meter = mp.Meter(scopeName)
```
Comment (bogdandrutu) [inline service/internal/proctelemetry/process_telemetry.go:84]: Do we need to keep a reference to the meter? Can you just pass it to `recordWithOtel`?

File: service/internal/proctelemetry/process_telemetry.go [hunk 9]
```diff
@@ -46,19 +66,31 @@ type processMetrics struct {
...
 // RegisterProcessMetrics creates a new set of processMetrics (mem, cpu) that can be used to measure
 // basic information about this process.
-func RegisterProcessMetrics(registry *metric.Registry, ballastSizeBytes uint64) error {
+func RegisterProcessMetrics(ocRegistry *metric.Registry, mp otelmetric.MeterProvider, useOtel bool, ballastSizeBytes uint64) error {
+	var err error
 	pm := &processMetrics{
 		startTimeUnixNano: time.Now().UnixNano(),
 		ballastSizeBytes:  ballastSizeBytes,
 		ms:                &runtime.MemStats{},
+		useOtelForMetrics: useOtel,
```
Comment (bogdandrutu) [inline service/internal/proctelemetry/process_telemetry.go:75]: No need to store it in the struct.

File: service/internal/proctelemetry/process_telemetry.go [hunk 10]
```diff
@@ -127,6 +159,102 @@ func RegisterProcessMetrics(registry *metric.Registry, ballastSizeBytes uint64)
...
+	}
+
+	pm.otelSysMem, err = pm.meter.Int64ObservableGauge(
+		"process_runtime_total_sys_memory_bytes",
+		instrument.WithDescription("Total bytes of memory obtained from the OS (see 'go doc runtime.MemStats.Sys')"),
+		instrument.WithUnit(unit.Bytes))
+	if err != nil {
+		return err
+	}
+	_, err = pm.meter.RegisterCallback(func(_ context.Context, o otelmetric.Observer) error {
```
Comment (bogdandrutu) [inline service/internal/proctelemetry/process_telemetry.go:217]: Why not using https://pkg.go.dev/go.opentelemetry.io/otel/metric@v0.35.0/instrument#WithInt64Callback in the previous call.

File: service/internal/proctelemetry/process_telemetry.go (hunk 8 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry.go]: Thanks for pointing this out! Fixed!

File: service/internal/proctelemetry/process_telemetry.go (hunk 9 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry.go]: Fixed!

File: service/internal/proctelemetry/process_telemetry.go (hunk 10 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry.go]: Didn't notice this option, but agreed this makes it cleaner. Thanks for pointing it out!

File: .chloggen/proctelemetry-processmetrics.yaml (hunk 7 above)
Comment (moh-osman3) [reply .chloggen/proctelemetry-processmetrics.yaml]: Fixed!

**Comment by moh-osman3** [comment]
@bogdandrutu @codeboten I rebased my change but noticed the contrib tests are now failing. Does the contrib repo need to be updated or is there a recent commit in the last 15 hours that is causing this failure? I don't see any references to `proctelemetry` in the contrib repo at all, so I'm a little confused here.

File: service/internal/proctelemetry/process_telemetry.go [hunk 11]
```diff
@@ -46,19 +63,30 @@ type processMetrics struct {
...
+
 	pm.proc, err = process.NewProcess(int32(os.Getpid()))
 	if err != nil {
 		return err
 	}
 
-	pm.processUptime, err = registry.AddFloat64DerivedCumulative(
+	if useOtel {
+		meter := mp.Meter(scopeName)
+		return pm.recordWithOtel(meter)
```
Comment (bogdandrutu) [inline service/internal/proctelemetry/process_telemetry.go:81]:
```suggestion
		return pm.recordWithOtel(mp.Meter(scopeName))
```

File: service/internal/proctelemetry/process_telemetry.go [hunk 12]
```diff
@@ -127,6 +155,84 @@ func RegisterProcessMetrics(registry *metric.Registry, ballastSizeBytes uint64)
...
+		"process_uptime",
+		instrument.WithDescription("Uptime of the process"),
+		instrument.WithUnit(unit.Unit("s")),
+		instrument.WithFloat64Callback(func(_ context.Context, o instrument.Float64Observer) error {
+			o.Observe(pm.updateProcessUptime())
+			return nil
+		}))
+	if err != nil {
+		return err
+	}
```
Comment (bogdandrutu) [inline service/internal/proctelemetry/process_telemetry.go:171]: Instead of checking and returning after first error, I think will be easier to read and probably a bit more helpful to aggregate all errors and return the result:

```suggestion
	var errs, err error

	pm.otelProcessUptime, err = meter.Float64ObservableCounter(
		"process_uptime",
		instrument.WithDescription("Uptime of the process"),
		instrument.WithUnit(unit.Unit("s")),
		instrument.WithFloat64Callback(func(_ context.Context, o instrument.Float64Observer) error {
			o.Observe(pm.updateProcessUptime())
			return nil
		}))
	errs = multierr.Append(errs, err)
```

**Review by bogdandrutu — APPROVED** [review]
